### PR TITLE
RNApuzzler layout

### PR DESCRIPTION
**Pull request by everyday847**
Enables layouts for large puzzles via the RNApuzzler algorithm. (Odd name! Whatever!)

Some food for thought:

 - [x] Right now I have it turn on automatically at 1000 nt. I am willing to be convinced of other thresholds, or for it to be activated on a per-puzzle basis or via a hotkey.
 - [x] In principle this works alongside `customLayout`, but it will not do collision detection vs. `customLayout` junctions. That is -- the "partial `customLayout`" will snap into place for relevant junctions, but the remaining `puzzlerLayout` junctions may overlap. This isn't a solvable issue because we can't feed their algorithm "fixed" domains.

### CODE REVIEW

**Comment by everyday847** [comment]
@luxaritas I'm happy to clean up a few logging statements and then get this live for broader testing. What do you think? I think usability/qol improvements (like a button) can come subsequently.

**Comment by everyday847** [comment]
I'm a fan of expediting features that make significant changes (like adding something new to the emscripten-built libraries; adding a bunch of new files); especially because I want players stress-testing the layout. I'll file some issues now.

File: src/eterna/EternaApp.ts [hunk 1]
```diff
@@ -150,7 +153,10 @@ export default class EternaApp extends FlashbangApp {
                 // We can only do this now, since we need the username and UID to connect
                 Eterna.chat = new ChatManager(this._params.chatboxID, Eterna.settings);
                 this.setLoadingText('Loading game...', null);
-                return Promise.all([this.initFoldingEngines(), TextureUtil.load(Bitmaps.all), Fonts.loadFonts()]);
+                return Promise.all([this.initFoldingEngines(),
+                    this.initLayoutEngines(),
+                    TextureUtil.load(Bitmaps.all),
+                    Fonts.loadFonts()]);
```
Comment (luxaritas) [inline src/eterna/EternaApp.ts:159]:
```suggestion
                return Promise.all([
                    this.initFoldingEngines(),
                    this.initLayoutEngines(),
                    TextureUtil.load(Bitmaps.all),
                    Fonts.loadFonts()
                ]);
```

File: src/eterna/emscripten/EmscriptenUtil.ts [hunk 2]
```diff
@@ -36,4 +36,14 @@ export default class EmscriptenUtil {
...
+
+    // /** Converts a C++ std::vector<std::vector<T>> with inner size 2 to a [T, T][] */
+    // public static arrayToStdVector<T>(array: T[]): stdcpp.vector<T> {
+    //     let n = array.length;
+    //     let vector: stdcpp.vector<T> = new stdcpp.vector<T>;
+    //     for (let ii = 0; ii < n; ++ii) {
+    //         vector.push_back(array[ii]);
+    //     }
+    //     return vector
+    // }
```
Comment (luxaritas) [inline src/eterna/emscripten/EmscriptenUtil.ts:48]: Remove?

File: src/eterna/layout/RNApuzzler.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        for (let ii = 0; ii < Xs.length; ++ii) {
+            zip.push([Xs[ii], Ys[ii]]);
+        }
+        return zip;
+    }
+
+    private constructor(lib: RNApuzzlerLib) {
+        super();
+        this._lib = lib;
+    }
```
Comment (luxaritas) [inline src/eterna/layout/RNApuzzler.ts:52]: Constructor should go right under create()

File: src/eterna/mode/PoseEdit/PoseEditMode.ts [hunk 4]
```diff
@@ -1232,6 +1232,10 @@ export default class PoseEditMode extends GameMode {
             } else if (!ctrl && key === KeyCode.Comma) {
                 Eterna.settings.simpleGraphics.value = !Eterna.settings.simpleGraphics.value;
                 handled = true;
+            } else if (!ctrl && key === KeyCode.KeyL) {
+                Eterna.settings.usePuzzlerLayout.value = !Eterna.settings.usePuzzlerLayout.value;
+                console.log('Setting for usePuzzlerLayout now', Eterna.settings.usePuzzlerLayout.value);
```
Comment (luxaritas) [inline src/eterna/mode/PoseEdit/PoseEditMode.ts:1237]: Can we remove this?

File: src/eterna/emscripten/EmscriptenUtil.ts (hunk 2 above)
Comment (everyday847) [reply src/eterna/emscripten/EmscriptenUtil.ts:48]: Sure.

**Pull request merged**